pkg/server/endpoints/eventTracker.go

```diff
@@ -0,0 +1,272 @@
+	}
+	if 0 < len(boundaries) {
+		boundaries[pollPeriods-1] = struct{}{}
+	}
+
+	boundaryList := slices.Collect(maps.Keys(boundaries))
+	slices.Sort(boundaryList)
+	if boundaryList == nil {
+		boundaryList = []uint{}
+	}
+
+	return boundaryList
+}
+
+func NewEventTracker(pollPeriods uint, boundaries []uint) *eventTracker {
+	if pollPeriods < 1 {
+		pollPeriods = 1
+	}
+
+	// cleanup boundaries into incrasing slice of no duplicates
+	boundaryMap := make(map[uint]bool)
+	filteredBounds := []uint{}
+	for _, boundary := range boundaries {
+		// trim duplicates and boundaries outside of polling range
+		if _, found := boundaryMap[boundary]; !found && boundary < pollPeriods {
+			boundaryMap[boundary] = true
+			filteredBounds = append(filteredBounds, boundary)
+		}
+	}
+	slices.Sort(filteredBounds)
+
+	initialPolls := uint(0)
+	switch {
+	case boundaries == nil, len(filteredBounds) == 0:
+		initialPolls = pollPeriods
+	default:
+		initialPolls = filteredBounds[0]
+	}
+
+	return &eventTracker{
+		initialPolls: initialPolls,
+		pollPeriods:  pollPeriods,
+		boundaries:   filteredBounds,
+		events:       make(map[uint]*eventStats),
+		pool: sync.Pool{
+			New: func() any {
+				return []uint(nil)
+			},
+		},
+	}
+}
+
+func (et *eventTracker) PollPeriods() uint {
+	return et.pollPeriods
+}
+
+func (et *eventTracker) InitialPolls() uint {
+	return et.initialPolls
+}
+
+func (et *eventTracker) PollBoundaries() []uint {
+	return et.boundaries
+}
+
+func (et *eventTracker) Polls() uint {
+	return et.initialPolls + uint(len(et.boundaries))
+}
+
+/**
+ * Starts tracking an event's walk through the boundaries.
+ */
+func (et *eventTracker) StartTracking(event uint) {
+	et.events[event] = &eventStats{
+		hash:  hash(event),
+		ticks: 0,
+		polls: 0,
+	}
+}
+
+/**
+ * Remove an event from the tracker.
+ *
+ * Events not explicitly removed will remove themselves
+ * after SelectEvents has been called PollPeriods() number
+ * of times.
+ */
+func (et *eventTracker) StopTracking(event uint) {
+	delete(et.events, event)
+}
+
+/**
+ * Selects the events one should pool for the next poll cycle.
+ *
+ * This algorithm determines if the events should be polled, and
+ * increments each event's time, allowing every event to act on
+ * the time the event was inserted into eventTracker.
+ *
+ * The event's boundary Index is computed, and if it is below the
+ * number of initial polls, the event is polled without further
+ * analysis.
+ *
+ * If the boundary index is inside a defined boundary, the width
+ * of the boundary is computed and the event's position within
+ * the virtual hash table is computed from "hash(event) % width".
+ * As the hash and width are stable, an event will always remain
+ * in the same slot within a boundary.
+ *
+ * If the event's ticks (the events local sense of time) match
+ * the event's slot within the boundary, the event is added to
+ * the poll list.
+ */
+func (et *eventTracker) SelectEvents() []uint {
+	pollList := et.pool.Get().([]uint)
```

Comment:
This isn't sufficient. In order to leverage the pool, memory needs to be returned to it. If you look at the pq branch, you'll see an additional "FreeEvents" method added to the tracker which is invoked by the caller of SelectEvents when it has finished to "return" the events slice back to the event tracker for reuse.

Reply:
That method is only called in benchmarks in the pq branch. FWIW.